File: src/main/java/link/infra/packwiz/installer/request/handlers/RequestHandlerZip.java
package link.infra.packwiz.installer.request.handlers;

import okio.Buffer;
import okio.BufferedSource;
import okio.Okio;
import okio.Source;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.locks.ReentrantLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;
import java.util.function.Predicate;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public abstract class RequestHandlerZip extends RequestHandlerHTTP {
	
	private final boolean modeHasFolder;
	
	public RequestHandlerZip(boolean modeHasFolder) {
		this.modeHasFolder = modeHasFolder;
	}
	
	private String removeFolder(String name) {
		if (modeHasFolder) {
			return name.substring(name.indexOf("/")+1);
		} else {
			return name;
		}
	}
	
	private class ZipReader {
		
		private final ZipInputStream zis;
		private final Map<URI, Buffer> readFiles = new HashMap<>();
		// Write lock implies access to ZipInputStream - only 1 thread must read at a time!
		final ReentrantLock filesLock = new ReentrantLock();
		private ZipEntry entry;

		private final BufferedSource zipSource;

		ZipReader(Source zip) {
			zis = new ZipInputStream(Okio.buffer(zip).inputStream());
			zipSource = Okio.buffer(Okio.source(zis));
		}
		
		// File lock must be obtained before calling this function
		private Buffer readCurrFile() throws IOException {
			Buffer fileBuffer = new Buffer();
			zipSource.readFully(fileBuffer, entry.getSize());
			return fileBuffer;
		}
		
		// File lock must be obtained before calling this function
		private Buffer findFile(URI loc) throws IOException, URISyntaxException {
			while (true) {
				entry = zis.getNextEntry();
				if (entry == null) {
					return null;
				}
				Buffer data = readCurrFile();
				URI fileLoc = new URI(removeFolder(entry.getName()));
				if (loc.equals(fileLoc)) {
					return data;
				} else {
					readFiles.put(fileLoc, data);
				}
			}
		}
		
		Source getFileSource(URI loc) throws Exception {
			filesLock.lock();
			// Assume files are only read once, allow GC by removing
			Buffer file = readFiles.remove(loc);
			if (file != null) {
				filesLock.unlock();
				return file;
			}
			
			file = findFile(loc);
			filesLock.unlock();
			return file;
		}
		
		URI findInZip(Predicate<URI> matches) throws Exception {
			filesLock.lock();
			for (URI file : readFiles.keySet()) {
				if (matches.test(file)) {
					filesLock.unlock();
					return file;
				}
			}
			
			while (true) {
				entry = zis.getNextEntry();
				if (entry == null) {
					filesLock.unlock();
					return null;
				}
				Buffer data = readCurrFile();
				URI fileLoc = new URI(removeFolder(entry.getName()));
				readFiles.put(fileLoc, data);
				if (matches.test(fileLoc)) {
					filesLock.unlock();
					return fileLoc;
				}
			}
		}
		
	}
	
	private final Map<URI, ZipReader> cache = new HashMap<>();
	private final ReentrantReadWriteLock cacheLock = new ReentrantReadWriteLock();
	
	protected abstract URI getZipUri(URI loc) throws Exception;
	
	protected abstract URI getLocationInZip(URI loc) throws Exception;
	
	@Override
	public abstract boolean matchesHandler(URI loc);

	@Override
	public Source getFileSource(URI loc) throws Exception {
		URI zipUri = getZipUri(loc);
		cacheLock.readLock().lock();
		ZipReader zr = cache.get(zipUri);
		cacheLock.readLock().unlock();
		if (zr == null) {
			cacheLock.writeLock().lock();
			// Recheck, because unlocking read lock allows another thread to modify it
			zr = cache.get(zipUri);
			if (zr == null) {
				Source src = super.getFileSource(zipUri);
				if (src == null) {
					cacheLock.writeLock().unlock();
					return null;
				}
				zr = new ZipReader(src);
				cache.put(zipUri, zr);
			}
			cacheLock.writeLock().unlock();
		}
		
		return zr.getFileSource(getLocationInZip(loc));
	}
	
	protected URI findInZip(URI loc, Predicate<URI> matches) throws Exception {
		URI zipUri = getZipUri(loc);
		cacheLock.readLock().lock();
		ZipReader zr = cache.get(zipUri);
		cacheLock.readLock().unlock();
		if (zr == null) {
			cacheLock.writeLock().lock();
			// Recheck, because unlocking read lock allows another thread to modify it
			zr = cache.get(zipUri);
			if (zr == null) {
				zr = new ZipReader(super.getFileSource(zipUri));
				cache.put(zipUri, zr);
			}
			cacheLock.writeLock().unlock();
		}
		
		return zr.findInZip(matches);
	}

}

File: src/main/kotlin/link/infra/packwiz/installer/request/handlers/RequestHandlerZip.kt
package link.infra.packwiz.installer.request.handlers

import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import okio.Buffer
import okio.Source
import okio.buffer
import okio.source
import java.util.*
import java.util.concurrent.locks.ReentrantLock
import java.util.concurrent.locks.ReentrantReadWriteLock
import java.util.function.Predicate
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import kotlin.concurrent.read
import kotlin.concurrent.withLock
import kotlin.concurrent.write

abstract class RequestHandlerZip(private val modeHasFolder: Boolean) : RequestHandlerHTTP() {
	private fun removeFolder(name: String): String {
		return if (modeHasFolder) {
			// TODO: replace with proper path checks once switched to Path??
			name.substring(name.indexOf("/") + 1)
		} else {
			name
		}
	}
	private fun removeFolder(path: HttpUrl): HttpUrl {
		return if (modeHasFolder) {
			path.newBuilder().removePathSegment(path.pathSegments.size - 1).build()
		} else {
			path
		}
	}

	private inner class ZipReader(zip: Source) {
		private val zis = ZipInputStream(zip.buffer().inputStream())
		private val readFiles: MutableMap<HttpUrl, Buffer> = HashMap()
		// Write lock implies access to ZipInputStream - only 1 thread must read at a time!
		val filesLock = ReentrantLock()
		private var entry: ZipEntry? = null

		private val zipSource = zis.source().buffer()

		// File lock must be obtained before calling this function
		private fun readCurrFile(): Buffer {
			val fileBuffer = Buffer()
			zipSource.readFully(fileBuffer, entry!!.size)
			return fileBuffer
		}

		// File lock must be obtained before calling this function
		private fun findFile(loc: HttpUrl): Buffer? {
			while (true) {
				entry = zis.nextEntry
				entry?.also {
					val data = readCurrFile()
					val fileLoc = removeFolder(it.name).toHttpUrl()
					if (loc == fileLoc) {
						return data
					} else {
						readFiles[fileLoc] = data
					}
				} ?: return null
			}
		}

		fun getFileSource(loc: HttpUrl): Source? {
			filesLock.withLock {
				// Assume files are only read once, allow GC by removing
				readFiles.remove(loc)?.also { return it }
				return findFile(loc)
			}
		}

		fun findInZip(matches: Predicate<HttpUrl>): HttpUrl? {
			filesLock.withLock {
				readFiles.keys.find { matches.test(it) }?.let { return it }

				do {
					val entry = zis.nextEntry?.also {
						val data = readCurrFile()
						val fileLoc = removeFolder(it.name).toHttpUrl()
						readFiles[fileLoc] = data
						if (matches.test(fileLoc)) {
							return fileLoc
						}
					}
				} while (entry != null)
				return null
			}
		}
	}

	private val cache: MutableMap<HttpUrl, ZipReader> = HashMap()
	private val cacheLock = ReentrantReadWriteLock()

	protected abstract fun getZipUri(loc: HttpUrl): HttpUrl
	protected abstract fun getLocationInZip(loc: HttpUrl): HttpUrl
	abstract override fun matchesHandler(loc: HttpUrl): Boolean

	override fun getFileSource(loc: HttpUrl): Source? {
		val zipUri = getZipUri(loc)
		var zr = cacheLock.read { cache[zipUri] }
		if (zr == null) {
			cacheLock.write {
				// Recheck, because unlocking read lock allows another thread to modify it
				zr = cache[zipUri]

				if (zr == null) {
					val src = super.getFileSource(zipUri) ?: return null
					zr = ZipReader(src).also { cache[zipUri] = it }
				}
			}
		}
		return zr?.getFileSource(getLocationInZip(loc))
	}

	protected fun findInZip(loc: HttpUrl, matches: Predicate<HttpUrl>): HttpUrl? {
		val zipUri = getZipUri(loc)
		return (cacheLock.read { cache[zipUri] } ?: cacheLock.write {
			// Recheck, because unlocking read lock allows another thread to modify it
			cache[zipUri] ?: run {
				// Create the ZipReader if it doesn't exist, return null if getFileSource returns null
				super.getFileSource(zipUri)?.let { ZipReader(it) }
						?.also { cache[zipUri] = it }
			}
		})?.findInZip(matches)
	}

}